### ISSUE TITLE
wasm-pack exceeds the maximum call stack size when parsing a "name" section that contains very long function names

### ISSUE THREAD

**Issue by RSSchermer** [rustwasm/wasm-bindgen#2567]
### Describe the Bug

Running a wasm-bindgen application with `npm run serve` results in the following error:

```
Failed to decode custom "name" section @8400654; ignoring (Maximum call stack size exceeded).
✖ ｢wdm｣: assets by status 1010 KiB [cached] 3 assets
assets by path . 114 KiB
  asset pkg_index_js.index.js 111 KiB [emitted]
  asset c9e4581f1d76676871ee.module.wasm 1.41 KiB [emitted] [immutable]
  asset index.html 1.31 KiB [emitted]
cached modules 962 KiB (javascript) 6.89 KiB (runtime) [cached] 49 modules
modules by layer 48.2 KiB (javascript) 9.26 MiB (webassembly)
  ./pkg/index.js 97 bytes [built]
  ./pkg/index_bg.js 48.1 KiB [built] [code generated]
  ./pkg/index_bg.wasm 40 bytes (javascript) 9.26 MiB (webassembly) [built] [code generated] [1 error]

ERROR in ./pkg/index_bg.wasm
Module parse failed: Unexpected section: 0x1e
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
Error: Unexpected section: 0x1e
    at new CompileError (/home/roland/Workspace/meshmerize-editor/node_modules/@webassemblyjs/helper-api-error/lib/index.js:40:109)
    at parseSection (/home/roland/Workspace/meshmerize-editor/node_modules/@webassemblyjs/wasm-parser/lib/decoder.js:1695:11)
    at Object.decode (/home/roland/Workspace/meshmerize-editor/node_modules/@webassemblyjs/wasm-parser/lib/decoder.js:1713:25)
    at decode (/home/roland/Workspace/meshmerize-editor/node_modules/@webassemblyjs/wasm-parser/lib/index.js:248:21)
    at WebAssemblyParser.parse (/home/roland/Workspace/meshmerize-editor/node_modules/webpack/lib/wasm-async/AsyncWebAssemblyParser.js:48:19)
    at /home/roland/Workspace/meshmerize-editor/node_modules/webpack/lib/NormalModule.js:1015:26
    at processResult (/home/roland/Workspace/meshmerize-editor/node_modules/webpack/lib/NormalModule.js:742:11)
    at /home/roland/Workspace/meshmerize-editor/node_modules/webpack/lib/NormalModule.js:806:5
    at /home/roland/Workspace/meshmerize-editor/node_modules/loader-runner/lib/LoaderRunner.js:406:3
    at iterateNormalLoaders (/home/roland/Workspace/meshmerize-editor/node_modules/loader-runner/lib/LoaderRunner.js:232:10)
 @ ./pkg/index.js 1:0-40 3:0-21
 @ ./index.js 3:0-15

webpack 5.37.1 compiled with 1 error in 21962 ms
ℹ ｢wdm｣: Failed to compile.
```

I did some "println debugging" by modifying the [parseNameSectionFunctions](https://github.com/xtuc/webassemblyjs/blob/0fb788b1b7785f2dcbb1b2e03b6564f607b4358c/packages/wasm-parser/src/decoder.js#L1184) to include `console.log(name)`. This allows me to watch it happily chug through the function names in the "name" section, until it encounters this bad boy:

```
core::ptr::drop_in_place<web_glitz::task::maybe_done::MaybeDone<web_glitz::task::join::Join<web_glitz::task::join::Join<web_glitz::task::join::JoinIter<alloc::boxed::Box<dyn web_glitz::task::gpu_task::GpuTask<web_glitz::runtime::rendering_context::Connection>+Output = ()>,web_glitz::runtime::rendering_context::Connection>,web_glitz::task::option_task::OptionTask<web_glitz::task::sequence::Sequence<web_glitz::task::map::Map<web_glitz::task::option_task::OptionTask<web_glitz::task::sequence::Sequence<web_glitz::buffer::UploadCommand<meshmerize_editor::rendering::simple_forward::active_mesh::painter::Uniforms,meshmerize_editor::rendering::simple_forward::active_mesh::painter::Uniforms>,web_glitz::rendering::render_pass::RenderPass<web_glitz::task::map::Map<web_glitz::task::join::Join<web_glitz::rendering::framebuffer::PipelineTask<web_glitz::task::join::JoinIter<web_glitz::task::map::Map<web_glitz::task::sequence::Sequence<web_glitz::task::sequence::Sequence<web_glitz::task::sequence::Sequence<web_glitz::task::sequence::Sequence<web_glitz::task::gpu_task::Empty,web_glitz::rendering::framebuffer::BindVertexBuffersCommand,web_glitz::rendering::framebuffer::PipelineTaskContext>,web_glitz::rendering::framebuffer::BindIndexBufferCommand,web_glitz::rendering::framebuffer::PipelineTaskContext>,web_glitz::rendering::framebuffer::BindResourcesCommand<[web_glitz::pipeline::resources::resource_bindings_encoding::BindGroupDescriptor; 2]>,web_glitz::rendering::framebuffer::PipelineTaskContext>,web_glitz::rendering::framebuffer::DrawIndexedCommand,web_glitz::rendering::framebuffer::PipelineTaskContext>,meshmerize_editor::rendering::simple_forward::active_mesh::painter::sculpt::SculptPainter<web_glitz::runtime::single_threaded::SingleThreadedContext>::paint::{{closure}}::{{closure}}::{{closure}}::{{closure}}>,web_glitz::rendering::framebuffer::PipelineTaskContext>>,web_glitz::task::option_task::OptionTask<web_glitz::rendering::framebuffer::PipelineTask<web_glitz::task::map::Map<web_glitz::task::sequence::Sequence<web_glitz::task::sequence::Sequence<web_glitz::task::sequence::Sequence<web_glitz::task::sequence::Sequence<web_glitz::task::gpu_task::Empty,web_glitz::rendering::framebuffer::BindVertexBuffersCommand,web_glitz::rendering::framebuffer::PipelineTaskContext>,web_glitz::rendering::framebuffer::BindIndexBufferCommand,web_glitz::rendering::framebuffer::PipelineTaskContext>,web_glitz::rendering::framebuffer::BindResourcesCommand<[web_glitz::pipeline::resources::resource_bindings_encoding::BindGroupDescriptor; 2]>,web_glitz::rendering::framebuffer::PipelineTaskContext>,web_glitz::rendering::framebuffer::DrawIndexedCommand,web_glitz::rendering::framebuffer::PipelineTaskContext>,meshmerize_editor::rendering::simple_forward::active_mesh::painter::sculpt::SculptPainter<web_glitz::runtime::single_threaded::SingleThreadedContext>::paint::{{closure}}::{{closure}}::{{closure}}::{{closure}}>>>,web_glitz::rendering::render_pass::RenderPassContext>,meshmerize_editor::rendering::simple_forward::active_mesh::painter::sculpt::SculptPainter<web_glitz::runtime::single_threaded::SingleThreadedContext>::paint::{{closure}}::{{closure}}>>,web_glitz::runtime::rendering_context::Connection>>,meshmerize_editor::rendering::simple_forward::active_mesh::painter::sculpt::SculptPainter<web_glitz::runtime::single_threaded::SingleThreadedContext>::paint::{{closure}}>,web_glitz::task::map::Map<web_glitz::task::sequence::Sequence<web_glitz::task::map::Map<web_glitz::task::sequence::Sequence<web_glitz::rendering::render_pass::RenderPass<web_glitz::task::sequence::Sequence<web_glitz::rendering::framebuffer::PipelineTask<web_glitz::task::sequence::Sequence<web_glitz::task::sequence::Sequence<web_glitz::task::sequence::Sequence<web_glitz::task::gpu_task::Empty,web_glitz::rendering::framebuffer::BindVertexBuffersCommand,web_glitz::rendering::framebuffer::PipelineTaskContext>,web_glitz::rendering::framebuffer::BindResourcesCommand<[web_glitz::pipeline::resources::resource_bindings_encoding::BindGroupDescriptor; 2]>,web_glitz::rendering::framebuffer::PipelineTaskContext>,web_glitz::rendering::framebuffer::DrawCommand,web_glitz::rendering::framebuffer::PipelineTaskContext>>,web_glitz::rendering::framebuffer::PipelineTask<web_glitz::task::sequence::Sequence<web_glitz::task::sequence::Sequence<web_glitz::task::sequence::Sequence<web_glitz::task::gpu_task::Empty,web_glitz::rendering::framebuffer::BindVertexBuffersCommand,web_glitz::rendering::framebuffer::PipelineTaskContext>,web_glitz::rendering::framebuffer::BindResourcesCommand<[web_glitz::pipeline::resources::resource_bindings_encoding::BindGroupDescriptor; 2]>,web_glitz::rendering::framebuffer::PipelineTaskContext>,web_glitz::rendering::framebuffer::DrawCommand,web_glitz::rendering::framebuffer::PipelineTaskContext>>,web_glitz::rendering::render_pass::RenderPassContext>>,web_glitz::rendering::render_pass::RenderPass<web_glitz::rendering::framebuffer::PipelineTask<web_glitz::task::sequence::Sequence<web_glitz::task::sequence::Sequence<web_glitz::task::sequence::Sequence<web_glitz::task::gpu_task::Empty,web_glitz::rendering::framebuffer::BindVertexBuffersCommand,web_glitz::rendering::framebuffer::PipelineTaskContext>,web_glitz::rendering::framebuffer::BindResourcesCommand<[web_glitz::pipeline::resources::resource_bindings_encoding::BindGroupDescriptor; 2]>,web_glitz::rendering::framebuffer::PipelineTaskContext>,web_glitz::rendering::framebuffer::DrawCommand,web_glitz::rendering::framebuffer::PipelineTaskContext>>>,web_glitz::runtime::rendering_context::Connection>,mesh_surface_data_web_glitz::seam_correction::SeamCorrectorColorBufferFloat<web_glitz::runtime::single_threaded::SingleThreadedContext,web_glitz::image::format::R16F>::correction_task::{{closure}}>,web_glitz::task::map::Map<web_glitz::task::sequence::Sequence<web_glitz::task::join::JoinIter<web_glitz::rendering::render_pass::RenderPass<web_glitz::rendering::framebuffer::PipelineTask<web_glitz::task::map::Map<web_glitz::task::sequence::Sequence<web_glitz::task::sequence::Sequence<web_glitz::task::sequence::Sequence<web_glitz::task::sequence::Sequence<web_glitz::task::gpu_task::Empty,web_glitz::rendering::framebuffer::BindVertexBuffersCommand,web_glitz::rendering::framebuffer::PipelineTaskContext>,web_glitz::rendering::framebuffer::BindIndexBufferCommand,web_glitz::rendering::framebuffer::PipelineTaskContext>,web_glitz::rendering::framebuffer::BindResourcesCommand<[web_glitz::pipeline::resources::resource_bindings_encoding::BindGroupDescriptor; 2]>,web_glitz::rendering::framebuffer::PipelineTaskContext>,web_glitz::rendering::framebuffer::DrawIndexedCommand,web_glitz::rendering::framebuffer::PipelineTaskContext>,meshmerize_editor::rendering::simple_forward::active_mesh::displacement_gradient_deriver::EditModeDisplacementGradientDeriver<web_glitz::runtime::single_threaded::SingleThreadedContext>::derive::{{closure}}::{{closure}}::{{closure}}::{{closure}}>>>,web_glitz::runtime::rendering_context::Connection>,web_glitz::image::texture_2d_array::GenerateMipmapCommand,web_glitz::runtime::rendering_context::Connection>,meshmerize_editor::rendering::simple_forward::active_mesh::displacement_gradient_deriver::EditModeDisplacementGradientDeriver<web_glitz::runtime::single_threaded::SingleThreadedContext>::derive::{{closure}}>,web_glitz::runtime::rendering_context::Connection>,meshmerize_editor::rendering::simple_forward::active_mesh::active_mesh_state::ActiveMeshState<web_glitz::runtime::single_threaded::SingleThreadedContext>::update_displacement_derived_atlasses_edit_mode::{{closure}}>,web_glitz::runtime::rendering_context::Connection>>,web_glitz::runtime::rendering_context::Connection>,web_glitz::task::option_task::OptionTask<web_glitz::task::map::Map<web_glitz::task::option_task::OptionTask<web_glitz::task::join::JoinIter<web_glitz::task::map::Map<web_glitz::task::sequence::Sequence<web_glitz::buffer::UploadCommand<meshmerize_editor::rendering::simple_forward::active_mesh::edit_buffer_adder::Uniforms,meshmerize_editor::rendering::simple_forward::active_mesh::edit_buffer_adder::Uniforms>,web_glitz::rendering::render_pass::RenderPass<web_glitz::rendering::framebuffer::PipelineTask<web_glitz::task::sequence::Sequence<web_glitz::task::sequence::Sequence<web_glitz::task::sequence::Sequence<web_glitz::task::gpu_task::Empty,web_glitz::rendering::framebuffer::BindVertexBuffersCommand,web_glitz::rendering::framebuffer::PipelineTaskContext>,web_glitz::rendering::framebuffer::BindResourcesCommand<[web_glitz::pipeline::resources::resource_bindings_encoding::BindGroupDescriptor; 2]>,web_glitz::rendering::framebuffer::PipelineTaskContext>,web_glitz::rendering::framebuffer::DrawCommand,web_glitz::rendering::framebuffer::PipelineTaskContext>>>,web_glitz::runtime::rendering_context::Connection>,meshmerize_editor::rendering::simple_forward::active_mesh::edit_buffer_adder::EditBufferAdder<web_glitz::runtime::single_threaded::SingleThreadedContext>::combine<web_glitz::image::format::R16F>::{{closure}}::{{closure}}>,web_glitz::runtime::rendering_context::Connection>>,meshmerize_editor::rendering::simple_forward::active_mesh::painter::sculpt::SculptPainter<web_glitz::runtime::single_threaded::SingleThreadedContext>::end_stroke::{{closure}}>>,web_glitz::runtime::rendering_context::Connection>,(((),core::option::Option<((),())>),core::option::Option<()>),web_glitz::runtime::rendering_context::Connection>>::h5be015561d53a2f3
```

At which point it produces the error above. I fully concur that that is a ridiculous type name. It results from a monad-like combination of GPU commands; it is essentially a typed GPU command buffer. 

There are several other "GPU command buffer" names similar to this in the "name" section that don't result in the error, but they are somewhat smaller (this one amounts to 9.4KB when saved to a `.txt` file, the next largest amounts to 7.2KB).

I won't pretend to fully understand how this long name subsequently blows out the V8 stack, however, given the moment of failure and the abnormal size of the name, I do suspect that its size is the cause.

### Steps to Reproduce

Unfortunately I cannot publicly share the source of the project that produces this error. I could however share the `pkg` directory if that would help.

### Expected Behavior

Either for wasm-pack to be able to handle arbitrarily large function names without exceeding the maximum stack size, or for wasm-bindgen to heuristically truncate long function names.

### Actual Behavior

NA

### Additional Context

NA

**Comment by RSSchermer** [comment]
I did some diagnosis ex juvantibus on a fork with [a modified version of the demangle function that truncates long function names](https://github.com/RSSchermer/wasm-bindgen/commit/a4aa5e69e753827188d4cba2d8c2f2fc02efcf0d). If I make wasm-pack use this modified version of wasm-bindgen-cli then it will finish building without errors and the resulting application runs correctly. I think that supports the theory that very long function names  are problematic and that truncating long function names is an effective solution.

I'd be happy to submit that as a PR if truncating indeed seems to be the appropriate solution.

**Comment by alexcrichton** [comment]
This looks like it's not wasm-bindgen that's blowing the stack, but rather some JS tooling? If that's the case how come wasm-bindgen would be changed as opposed to fixing the bug in JS?

**Comment by RSSchermer** [comment]
As implied, I'm not at all sure if truncation is the appropriate solution. I definitely found it surprising that [@webassemblyjs/wasm-parser](https://github.com/xtuc/webassemblyjs/tree/master/packages/wasm-parser) cannot handle arbitrarily large names (I'll note that I also don't quite understand it's role in the toolchain).  On the other hand, I also found it somewhat surprising that individual names (resulting from imho reasonable code) are able to get to a size of 9.4KB.

It is not currently clear to me how the long name causes [parseNameSectionFunctions](https://github.com/xtuc/webassemblyjs/blob/0fb788b1b7785f2dcbb1b2e03b6564f607b4358c/packages/wasm-parser/src/decoder.js#L1184) to exceed the stack size, everything it does seems reasonable to me, but I'm not nearly an expert on V8's inner workings. I've opened [#922](https://github.com/xtuc/webassemblyjs/issues/922) on [xtuc/webassemblyjs](https://github.com/xtuc/webassemblyjs) in the hopes that perhaps the author has a deeper understanding.

In short, I don't believe I know enough to have a strong opinion on whether either, or perhaps even both solutions are appropriate.
